### PR TITLE
feat: Add MultiIndex subclass.

### PR DESCRIPTION
**Pull request by TrevorBergeron**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

### CODE REVIEW

File: bigframes/core/indexes/base.py [hunk 1]
```diff
@@ -42,14 +43,21 @@
...
-    def __init__(
+    # Overrided on __new__ to create subclasses like python does
+    def __new__(
         self,
         data=None,
         dtype=None,
         *,
         name=None,
         session=None,
+        linked_frame: Union[
```
[the remark this answers is not in the document]
Comment (TrevorBergeron) [reply bigframes/core/indexes/base.py]: This is for internal purposes only. Removed from constructor, now directly set the property after initialized.

File: bigframes/core/indexes/base.py [hunk 2]
```diff
@@ -454,24 +475,27 @@ def __len__(self):
         return self.shape[0]
 
 
-# Index that mutates the originating dataframe/series
-class FrameIndex(Index):
```
[the remark this answers is not in the document]
Comment (TrevorBergeron) [reply bigframes/core/indexes/base.py:458]: I am deleting a class yes, but it was more of an implementation detail of the Index class. Should only affect users if they were somehow depending on exact type or type name in their code, which would be very strange.

[the remark this answers is not in the document]
Comment (tswast) [reply bigframes/core/indexes/base.py:458]: Good catch @GarrettWu.

Given that we say `) -> Index:` in the only method that would return such an object and we don't explicitly document `FrameIndex`, I'm okay overlooking this and treating `FrameIndex` as a private implementation detail.

File: bigframes/core/indexes/base.py [hunk 3]
```diff
@@ -454,24 +475,27 @@ def __len__(self):
...
-    def __init__(
-        self,
-        series_or_dataframe: typing.Union[
-            bigframes.series.Series, bigframes.dataframe.DataFrame
-        ],
-    ):
-        super().__init__(series_or_dataframe._block)
-        self._whole_frame = series_or_dataframe
+class MultiIndex(Index, vendored_pandas_multindex.MultiIndex):
```
Comment (tswast) [inline bigframes/core/indexes/base.py:478]: As much as possible, I'd like our module hierarch to reflect pandas. In pandas, this is defined in `pandas/core/indexes/multi.py`. Anything under `core` is considered private-ish, but these paths do leak out in some of the pandas docs.

https://github.com/pandas-dev/pandas/blob/5232bee8df3b57766c44b62152aa3fdd24e40ada/pandas/core/indexes/multi.py#L224

Let's make a `bigframes/core/indexes/multi.py` file and put this there.

Comment (TrevorBergeron) [reply bigframes/core/indexes/base.py]: done

File: bigframes/core/indexes/base.py [hunk 4]
```diff
@@ -42,14 +43,21 @@
...
 class Index(vendored_pandas_index.Index):
     __doc__ = vendored_pandas_index.Index.__doc__
+    _query_job = None
+    _block: blocks.Block
+    _linked_frame: Union[bigframes.dataframe.DataFrame, bigframes.series.Series, None]
 
-    def __init__(
+    # Overrided on __new__ to create subclasses like python does
+    def __new__(
         self,
```
[the remark this answers is not in the document]
Comment (TrevorBergeron) [reply bigframes/core/indexes/base.py]: fixed

**Review by GarrettWu — APPROVED** [review]

**Pull request merged**